File: src/server/providers/validation.ts
import { CAPABILITIES, MEDIA_LIMITS } from "@/lib/capabilities";
import type { Network, ResolvedContent } from "@/lib/types";
import { ProviderError } from "./errors";

/**
 * Pre-publish validation: character limits, media counts, MIME types, sizes,
 * aspect/duration constraints, and unsupported combinations. Runs before any
 * external write. Provider-specific nuances live in each adapter's own
 * validate() which calls this first.
 */

export function validateContent(network: Network, content: ResolvedContent): void {
  const caps = CAPABILITIES[network];
  const fail = (message: string): never => {
    throw new ProviderError(network, message);
  };

  const text = content.text.trim();
  const images = content.media.filter((m) => m.kind === "image");
  const videos = content.media.filter((m) => m.kind === "video");

  // Text rules
  if (text.length > caps.maxChars) {
    fail(`${caps.network}: text is ${text.length} characters (limit ${caps.maxChars})`);
  }

  // Media presence rules
  if (caps.requiresMedia && content.media.length === 0) {
    fail(`${caps.network}: at least one image or video is required`);
  }
  if (!text && content.media.length === 0) {
    fail(`${caps.network}: post has no text and no media`);
  }

  // Counts
  if (images.length > 0 && videos.length > 0 && !caps.allowsMixedMedia) {
    fail(`${caps.network}: cannot mix images and video in one post`);
  }
  if (images.length > caps.maxImages) {
    fail(`${caps.network}: ${images.length} images selected (limit ${caps.maxImages})`);
  }
  if (videos.length > 1) {
    fail(`${caps.network}: only one video per post`);
  }
  if (videos.length > 0 && !caps.allowsVideo) {
    fail(`${caps.network}: video is not supported`);
  }
  if (images.length > 1 && !caps.allowsCarousel && !caps.allowsMixedMedia) {
    // e.g. X/Bluesky allow up to 4 images but not as a "carousel" container;
    // handled by maxImages above. This branch is for future providers.
  }

  // Per-item checks
  for (const item of content.media) {
    const limits = MEDIA_LIMITS[item.kind];
    const maxBytes =
      item.kind === "image" && !caps.autoOptimizeImages
        ? Math.min(limits.maxBytes, caps.maxImageBytes ?? limits.maxBytes)
        : limits.maxBytes;
    if (!(limits.mimeTypes as readonly string[]).includes(item.mimeType)) {
      fail(`${caps.network}: unsupported ${item.kind} type ${item.mimeType}`);
    }
    if (item.size > maxBytes) {
      fail(
        `${caps.network}: ${item.name} exceeds the ${Math.round(maxBytes / 1_000_000)}MB ${item.kind} limit`,
      );
    }
    if (item.kind === "video") {
      const videoLimits = MEDIA_LIMITS.video;
      if (
        item.durationSeconds !== undefined &&
        item.durationSeconds > videoLimits.maxDurationSeconds
      ) {
        fail(`${caps.network}: video is longer than ${videoLimits.maxDurationSeconds}s`);
      }
    }
    if (item.kind === "image" && item.width && item.height) {
      const ratio = item.width / item.height;
      if (ratio < 1 / 20 || ratio > 20) {
        fail(`${caps.network}: image ${item.name} has an unsupported aspect ratio`);
      }
    }
  }
}

File: Blather/Domain/Validation.swift
import Foundation

struct ProviderError: Error, LocalizedError {
    var network: Network
    var message: String
    var retryable: Bool
    var status: Int?

    init(network: Network, _ message: String, retryable: Bool = false, status: Int? = nil) {
        self.network = network
        self.message = message
        self.retryable = retryable
        self.status = status
    }

    var errorDescription: String? { message }
}

enum ContentValidation {
    static func validate(_ network: Network, content: ResolvedContent) throws {
        let caps = Capabilities.capabilities(for: network)
        let text = content.text.trimmingCharacters(in: .whitespacesAndNewlines)
        let images = content.media.filter { $0.kind == .image }
        let videos = content.media.filter { $0.kind == .video }

        if text.count > caps.maxChars {
            throw ProviderError(network: network, "\(caps.network.rawValue): text is \(text.count) characters (limit \(caps.maxChars))")
        }
        if caps.requiresMedia, content.media.isEmpty {
            throw ProviderError(network: network, "\(caps.network.rawValue): at least one image or video is required")
        }
        if text.isEmpty, content.media.isEmpty {
            throw ProviderError(network: network, "\(caps.network.rawValue): post has no text and no media")
        }
        if !images.isEmpty, !videos.isEmpty, !caps.allowsMixedMedia {
            throw ProviderError(network: network, "\(caps.network.rawValue): cannot mix images and video in one post")
        }
        if images.count > caps.maxImages {
            throw ProviderError(network: network, "\(caps.network.rawValue): \(images.count) images selected (limit \(caps.maxImages))")
        }
        if videos.count > 1 {
            throw ProviderError(network: network, "\(caps.network.rawValue): only one video per post")
        }
        if !videos.isEmpty, !caps.allowsVideo {
            throw ProviderError(network: network, "\(caps.network.rawValue): video is not supported")
        }

        for item in content.media {
            let mimeOk: Bool
            let maxBytes: Int
            if item.kind == .image {
                mimeOk = MediaLimits.imageMimeTypes.contains(item.mimeType)
                if caps.autoOptimizeImages {
                    maxBytes = MediaLimits.imageMaxBytes
                } else {
                    maxBytes = min(MediaLimits.imageMaxBytes, caps.maxImageBytes ?? MediaLimits.imageMaxBytes)
                }
            } else {
                mimeOk = MediaLimits.videoMimeTypes.contains(item.mimeType)
                maxBytes = MediaLimits.videoMaxBytes
            }
            if !mimeOk {
                throw ProviderError(network: network, "\(caps.network.rawValue): unsupported \(item.kind.rawValue) type \(item.mimeType)")
            }
            if item.size > maxBytes {
                let mb = Int((Double(maxBytes) / 1_000_000).rounded())
                throw ProviderError(network: network, "\(caps.network.rawValue): \(item.name) exceeds the \(mb)MB \(item.kind.rawValue) limit")
            }
            if item.kind == .video, let duration = item.durationSeconds, duration > MediaLimits.videoMaxDurationSeconds {
                throw ProviderError(network: network, "\(caps.network.rawValue): video is longer than \(Int(MediaLimits.videoMaxDurationSeconds))s")
            }
            if item.kind == .image, let width = item.width, let height = item.height, height > 0 {
                let ratio = Double(width) / Double(height)
                if ratio < 1.0 / 20.0 || ratio > 20 {
                    throw ProviderError(network: network, "\(caps.network.rawValue): image \(item.name) has an unsupported aspect ratio")
                }
            }
        }
    }
}
